### DISCUSSION TITLE
Maven P2 site usecase and pomless example

### DISCUSSION THREAD

**Discussion by jgsuess** [eclipse-tycho/tycho#430] [Q&A]
I am looking for a use case description and ideally example for the [tycho-p2-repository:assemble-maven-repository](https://www.eclipse.org/tycho/sitedocs/tycho-p2/tycho-p2-repository-plugin/assemble-maven-repository-mojo.html). My understanding is that this would allow publishing P2 equivalent information to an M2 repository, such that it can be consumed by PDE with M2E or tycho builds. If so, this would be a great simplification for our usecases in the company. Ideally I would like to know how this integrates with pomless builds.

Thanks for helping the noobie.

**Comment by laeubi** [comment]
The use-case is mainly intended for projects that are "plain-maven" (e.g. no tycho build) but still producing maven bundles (e.g bnd/felix-maven plugin) to allow them publish a P2 site without extra effort. As an example one could take a look at the [jetty-p2 module](https://github.com/eclipse/jetty.project/tree/jetty-10.0.x/jetty-p2).

The resulting site is deployed to a maven repository and could then be consumed from there, assuming the "plain-maven" artifacts are already deployed there.

You should also be able to use this within a tycho build given the following perquisites:

1. all your tycho-bundles are deployed to a maven artifact server
2. you create a module that is a plain maven project and either references all things to include as maven dependencies, or you take otherwise sure the module is run last in your build and use the reactor project approach.

> If so, this would be a great simplification for our usecases in the company

The maven p2 site mojo is relative new. So feel free to describe a bit more your use-case and we might be able to extend/enhance the mojo so it could be used in that scenario as well. I could imagine that for a tycho build it would be a good idea to have a separate packaging type that takes care of the reactor project order and alike.

**Comment by jgsuess** [comment]
In our company we use Gitlab and its project-associated [Maven Repositories](https://docs.gitlab.com/ee/user/packages/maven_repository/) for plain Java projects. This is great, because it allows us to store and verify everything in one go. For Tycho builds, we currently post our sites to cloud buckets as [Gitlab does not offer P2](https://gitlab.com/gitlab-org/gitlab/-/issues/36895). This creates a separate silo that needs to be managed, cleaned and is not amenable to standard tooling. P2 tooling is not great to start with.
So, if we could produce and deploy to Maven instead with a metadata capsule that contains the download sources, this would be an absolute game-changer for us. Ideally this would be wired in the root POM of a POM-less build. Then we could feed all our builds Maven only until we come to the point where we want to deploy the feature for the general public. Who knows, the Maven site might even make its way into Eclipse mainstream as an alternative format?

**Reply by laeubi** [reply]
I think the first step would then to setup your tycho build in a way that it deploys all bundles to the project-associated maven repository.
If that works there are two choices:

1. You can consume them directly with the maven-target-location feature (no p2site required)
2. You can publish the site as a maven-p2 site as well and consume this in tycho/eclipse like any regular site

As this is interesting for others as well just let us know how it works and where/how we can improve here. If possible you might want to share a runnable example that shows some restrictions or pitfalls so we can work on it directly.

If you want to try it out, the p2site of the jetty release could be added to a target with the following URLs:

Version 10: `mvn:org.eclipse.jetty:jetty-p2:10.0.7:zip:p2site`
Version 11: `mvn:org.eclipse.jetty:jetty-p2:11.0.7:zip:p2site`

@akurtakov might can elaborate a bit more if the jetty P2 site is actually used somewhere already inside platform code.

**Reply by jgsuess** [reply]
I have already added the plugin to a test build, but I am failing to locate the binary. I had expected it to appear next to the other site in my site build location. If it is deploying to the local M2 repository, then the impediment would be to distinguish the items produced automatically. I am not quite sure how to wire it up. Maybe I will make a minimal example and post it here for discussion. The downstream is trivial. I understand how I would consume things from this type of storage.

**Reply by laeubi** [reply]
The plugin produces a separate attached artifact, as mentioned above you have to create an own module for it with a pom (you can't use it in a pomless style as there is no equivalent PDE item  available for this kind of feature).

**Reply by laeubi** [reply]
> Maybe I will make a minimal example and post it here for discussion.

That would be great, it might also be possible to use `-Durl=file:/foldertodeploy` then to make some kind of integrationtest out of it.

**Reply by mickaelistria** [reply]
> Maven instead with a metadata capsule that contains the download sources, this would be an absolute game-changer for us.

Tycho also has https://www.eclipse.org/tycho/sitedocs/tycho-p2/tycho-p2-repository-plugin/remap-artifacts-to-m2-repo-mojo.html which allows to generate a p2 repository that does reference content from a Maven repository, instead of copying them. So the repository becomes deploy-able to Nexus and -in combination with [Nexus Unzip Plugin](https://wiki.eclipse.org/Tycho/Nexus_Unzip_Plugin), it can then be used by p2 to consume the same jars as what plain Maven would do.

> Ideally this would be wired in the root POM of a POM-less build. 

The repository would still be an artifact, you'd still need a category.xml or some dedicated pom configuration to generate the right content for the p2 repo.

> Then we could feed all our builds Maven only until we come to the point where we want to deploy the feature for the general public

I think you can already do that with what's in place in Tycho and PDE to consume Maven artifacts and p2-ify them during the build. Many OSGi plugins are build with an approach that doesn't involve p2 nor Tycho for most of the parts, and are then consumed by an Eclipse feature or RCP thanks to existing mechanisms that allows to consume those artifacts in a p2 world. This usually requires 2 build steps though, but it works fine.

One alternative way to consider the problem is to make it a "hosting" problem. Maven repos are often managed by Nexus; and Nexus knows all available artifacts; so Nexus could generate a p2 repository that describes all the content that's available for consumption in p2. Maybe a Nexus plugin is a better, more generic and more powerful solution than a new workflow inside the build, a workflow that would have to be replicated on all builds...

**Reply by laeubi** [reply]
> which allows to generate a p2 repository that does reference content from a Maven repository

The problem is that this requires to first build the repository and then clear the converted artifacts if one cares about deployment size. But the Unzip plugin is not necessary anymore as one could now use a mvn URL instead.

**Comment by laeubi** [comment]
I have created https://github.com/eclipse/tycho/issues/436 to investigate more convenient ways to use this inside a tycho build.

In the meantime you can use the following approach:

1. create a new folder `mvnsite` in the root of your build tree
2. add `<module>mvnsite</module>` to your root pom.xml **at the end of all reactor projects!** (as an alternative, you can add a maven dependency to your product build or something that is build after all plugins in your reactor to the following pom)
3. add the following pom to mvnsite, adjust to your needs:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>tycho.example</groupId>
	<artifactId>p2-site</artifactId>
	<version>1.0.0-SNAPSHOT</version>
	<packaging>pom</packaging>

	<build>
		<plugins>
			<plugin>
				<groupId>org.eclipse.tycho</groupId>
				<artifactId>tycho-p2-repository-plugin</artifactId>
				<version>2.5.0</version>
				<executions>
					<execution>
						<configuration>
							<categoryName>My Bundles</categoryName>
							<includeReactor>true</includeReactor>
							<includeDependencies>false</includeDependencies>
							<includePGPSignature>false</includePGPSignature>
						</configuration>
						<id>maven-p2-site</id>
						<phase>package</phase>
						<goals>
							<goal>assemble-maven-repository</goal>
						</goals>
					</execution>
				</executions>
			</plugin>
		</plugins>
	</build>

</project>
```

This will then create a repository under `target` folder that contains all your bundles in a way that could be deployed to maven repository and consumed from that.

**Reply by jgsuess** [reply]
Thanks! This is very useful already. I am now looking for the commands to perform the actual publication of thethe site. Then the prototypical workflow would be complete and we would be up and away!

**Reply by jgsuess** [reply]
For any reading above instructions, the POM needs to be located under `sites/mvnsite`. Only then is the module being built, the extra modules entry did not seem to be necessary for me.
